agent/structs/config_entry_status.go

```diff
@@ -195,3 +210,466 @@ func (u *StatusUpdater) UpdateEntry() (ControlledConfigEntry, bool) {
+	RouteReasonNoUpstreamServicesTargeted RouteConditionReason = "NoUpstreamServicesTargeted"
+
+	// This condition indicates whether the controller was able to resolve all
+	// the object references for the Route.
+	//
+	// Possible reasons for this condition to be true are:
+	//
+	// * "ResolvedRefs"
+	//
+	// Possible reasons for this condition to be false are:
+	//
+	// * "RefNotPermitted"
+	// * "InvalidKind"
+	// * "BackendNotFound"
+	//
+	// Controllers may raise this condition with other reasons,
+	// but should prefer to use the reasons listed above to improve
+	// interoperability.
+	RouteConditionResolvedRefs RouteConditionType = "ResolvedRefs"
+
+	// This reason is used with the "ResolvedRefs" condition when the condition
+	// is true.
+	RouteReasonResolvedRefs RouteConditionReason = "ResolvedRefs"
+
+	// This reason is used with the "ResolvedRefs" condition when
+	// one of the Listener's Routes has a BackendRef to an object in
+	// another namespace, where the object in the other namespace does
+	// not have a ReferenceGrant explicitly allowing the reference.
+	RouteReasonRefNotPermitted RouteConditionReason = "RefNotPermitted"
+
+	// This reason is used with the "ResolvedRefs" condition when
+	// one of the Route's rules has a reference to an unknown or unsupported
+	// Group and/or Kind.
+	RouteReasonInvalidKind RouteConditionReason = "InvalidKind"
+
+	// This reason is used with the "ResolvedRefs" condition when one of the
+	// Route's rules has a reference to a resource that does not exist.
+	RouteReasonBackendNotFound RouteConditionReason = "BackendNotFound"
+)
+
+// the following statuses are custom to Consul
+const (
+	// This condition indicates whether the route was able to successfully bind the
+	// Listener on the gateway
+	// Possible reasons for this condition to be true are:
+	//
+	// * "Bound"
+	//
+	// Possible reasons for this condition to be false are:
+	//
+	// * "FailedToBind"
+	// * "GatewayNotFound"
+	//
+	RouteConditionBound RouteConditionType = "Bound"
+
+	// This reason is used with the "Bound" condition when the condition
+	// is true
+	RouteReasonBound RouteConditionReason = "Bound"
+
+	// This reason is used with the "Bound" condition when the route failed
+	// to bind to the gateway
+	RouteReasonFailedToBind RouteConditionReason = "FailedToBind"
+
+	// This reason is used with the "Bound" condition when the route fails
+	// to find the gateway
+	RouteReasonGatewayNotFound RouteConditionReason = "GatewayNotFound"
+)
+
+// NewRouteCondition is a helper to build allowable Conditions for a Route config entry
+func NewRouteCondition(name RouteConditionType, status ConditionStatus, reason RouteConditionReason, message string, ref ResourceReference) Condition {
+	if err := checkRouteConditionReason(name, status, reason); err != nil {
+		// note we panic here because an invalid combination is a programmer error
+		// this  should never actually be hit
+		panic(err)
```

Comment:
we panic here and in the gateway constructor because an error here is 100% programmer error